projects/rocprofiler-systems/source/lib/core/trace_cache/rocpd_post_processing.cpp

```diff
@@ -375,6 +376,265 @@ rocpd_post_processing::get_pmc_event_with_sample_callback() const
+        insert_event_and_sample(
+            is_mem_usage_enabled, trait::name<category::amd_smi_memory_usage>::value,
+            info::annotate_with_device_id<category::amd_smi_memory_usage>(
+                _amd_smi.device_id)
+                .c_str(),
+            _amd_smi.mem_usage);
+
+        if(!is_vcn_enabled && !is_jpeg_enabled)
+        {
+            return;
+        }
+
+        std::vector<xcp_metrics_t> xcp_metrics;
+        bool                       is_vcn_activity_supported;
+        bool                       is_jpeg_activity_supported;
+        deserialize_xcp_metrics(_amd_smi.xcp_activity, is_vcn_activity_supported,
+                                is_jpeg_activity_supported, xcp_metrics);
+
+        auto insert_xcp_metrics = [&](auto category, bool _is_enabled,
+                                      const std::vector<uint16_t>& data,
+                                      std::optional<size_t>        _idx = std::nullopt) {
+            if(!_is_enabled)
+            {
+                return;
+            }
+
+            using Category = std::decay_t<decltype(category)>;
+
+            for(size_t clk = 0; clk < data.size(); ++clk)
+            {
+                const auto value = data[clk];
+                if(value == std::numeric_limits<uint16_t>::max())
+                {
+                    continue;
+                }
+
+                auto pmc_name   = info::annotate_category<Category>(_idx, clk);
+                auto track_name = info::annotate_with_device_id<Category>(
+                    _amd_smi.device_id, _idx, clk);
+
+                insert_event_and_sample(_is_enabled, pmc_name.c_str(), track_name.c_str(),
+                                        value);
+            }
+        };
+
+        for(size_t idx = 0; idx < xcp_metrics.size(); ++idx)
+        {
+            auto dimension =
+                xcp_metrics.size() == 1 ? std::nullopt : std::make_optional<size_t>(idx);
+
+            insert_xcp_metrics(category::amd_smi_vcn_activity{}, is_vcn_enabled,
+                               xcp_metrics[idx].vcn_busy, dimension);
+
+            insert_xcp_metrics(category::amd_smi_jpeg_activity{}, is_jpeg_enabled,
+                               xcp_metrics[idx].jpeg_busy, dimension);
+        }
+    };
+}
+
+postprocessing_callback
+rocpd_post_processing::get_cpu_freq_sample_callback() const
+{
+    struct core_freq_sample
+    {
+        size_t id;
+        float  value;
+    };
+
+    auto deserialize_freqs = [](std::vector<uint8_t>& data) {
+        std::vector<core_freq_sample> result;
+        size_t                        offset = 0;
+
+        while(offset + sizeof(float) + sizeof(size_t) <= data.size())
+        {
+            core_freq_sample core_sample;
+            core_sample.id = *reinterpret_cast<size_t*>(data.data() + offset);
+            offset += sizeof(size_t);
+            core_sample.value = *reinterpret_cast<float*>(data.data() + offset);
+            offset += sizeof(float);
```

Comment:
```suggestion
            std::memcpy(&core_sample.id, data.data() + offset, sizeof(size_t));
            offset += sizeof(size_t);
            std::memcpy(&core_sample.value, data.data() + offset, sizeof(float));
            offset += sizeof(float);
```
Probaby okay, but using memcpy may be safer.